go-controller/pkg/node/egressfirewall_dns/egressfirewall_dns.go

```diff
@@ -0,0 +1,271 @@
+//    entry is used
+// 2. If the defaultInterval has run (30 min) without updating the DNS server is manually queried
+func (e *EgressDNS) Run(defaultInterval time.Duration) {
+	var dnsNamespace dnsNamespace
+	var ttl time.Time
+	var timeSet bool
+	// initially the next DNS Query happens at the default interval
+	durationTillNextQuery := defaultInterval
+	go func() {
+		for {
+			// Wait for the given duration or until something gets added
+			select {
+			case dnsNamespace := <-e.added:
+				if err := e.dns.Add(dnsNamespace.dnsName); err != nil {
+					utilruntime.HandleError(err)
+				}
+				if err := e.updateEntryForName(dnsNamespace); err != nil {
+					utilruntime.HandleError(err)
+				}
+			case <-time.After(durationTillNextQuery):
+				if len(dnsNamespace.dnsName) > 0 {
+					if _, err := e.Update(dnsNamespace.dnsName); err != nil {
+						utilruntime.HandleError(err)
+					}
+					if err := e.updateEntryForName(dnsNamespace); err != nil {
+						utilruntime.HandleError(err)
+					}
+				}
+			case dnsName := <-e.removed:
+				// when there is a dnsName removed the query time should be updated
+				e.dns.Delete(dnsName)
+			case <-e.stopChan:
+				return
+			case <-e.controllerStop:
+				return
+			}
+
+			// before waiting on the signals get the next time this thread needs to wake up
+			ttl, dnsNamespace.dnsName, timeSet = e.dns.GetNextQueryTime()
+			if time.Until(ttl) > defaultInterval || !timeSet {
+				durationTillNextQuery = defaultInterval
+			} else {
+				durationTillNextQuery = time.Until(ttl)
+			}
+			dnsNamespace.namespace = ""
+		}
+	}()
+
+}
+
+func (e *EgressDNS) Remove(dnsNames []string, namespace string) error {
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	for _, dnsName := range dnsNames {
+		delete(e.dnsEntries[dnsName].namespaces, namespace)
+		if len(e.dnsEntries[dnsName].namespaces) == 0 {
+			delete(e.dnsEntries, dnsName)
+			e.signalRemoved(dnsName)
+			dnsObject, err := e.wf.GetDNSObject(e.nodeName)
+			if err != nil {
+				return err
+			}
+			if _, exists := dnsObject.Spec.DNSObjectEntries[dnsName]; exists {
+				delete(dnsObject.Spec.DNSObjectEntries, dnsName)
+				if len(dnsObject.Spec.DNSObjectEntries) == 0 {
+					klog.Infof("Deleteing DNSObject %s from the cluster", dnsObject.Name)
+					err := e.k.DeleteDNSObject(dnsObject.Name)
+					if err != nil {
+						return err
+					}
+
+				} else {
+					err := e.k.UpdateDNSObject(dnsObject)
+					if err != nil {
+						return err
+					}
+				}
+
+			}
+
+		}
+
+	}
+	return nil
+
+}
+
+func (e *EgressDNS) Shutdown() {
+	close(e.stopChan)
+}
+
+func (e *EgressDNS) signalRemoved(dnsName string) {
+	e.removed <- dnsName
+}
+
+func (e *EgressDNS) signalAdded(dnsNS dnsNamespace) {
+	e.added <- dnsNS
+}
+
+// *** Functions below are used for testing purposes only ***
+
+// returns the fields of a dnsEntry for a given name
+func (e *EgressDNS) GetDNSEntry(dnsName string) ([]net.IP, []string, bool) {
```

Comment:
Can these functions be in a `_test.go` file? Or are they needed in a different package?

Reply:
@squeed it needs to be there it is used by the test files to verify that the DNSEntry is correct and that test is in a different package